File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/object2d/Polygon.java
package pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d;

import com.google.common.base.Objects;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.render.*;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.util.math.Matrix4f;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasClient;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ColourableObject;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectRegistry;
import pw.switchcraft.plethora.util.ByteBufUtils;
import pw.switchcraft.plethora.util.Vec2d;

import javax.annotation.Nonnull;
import java.util.ArrayList;

public class Polygon extends ColourableObject implements MultiPointResizable2d {
	protected final ArrayList<Vec2d> points = new ArrayList<>();

	protected Polygon(int id, int parent, byte type) {
		super(id, parent, type);
	}

	public Polygon(int id, int parent) {
		super(id, parent, ObjectRegistry.POLYGON_2D);
	}

	@Nonnull
	@Override
	public Vec2d getPoint(int idx) {
		return points.get(idx);
	}

	@Override
	public void setVertex(int idx, @Nonnull Vec2d point) {
		if (!Objects.equal(points.get(idx), point)) {
			points.set(idx, point);
			setDirty();
		}
	}

	@Override
	public int getVertices() {
		return points.size();
	}

	@Override
	public void removePoint(int idx) {
		points.remove(idx);
		setDirty();
	}

	@Override
	public void addPoint(int idx, @Nonnull Vec2d point) {
		if (idx == points.size()) {
			points.add(point);
		} else {
			points.add(idx, point);
		}

		setDirty();
	}

	@Override
	public void writeInitial(@Nonnull PacketByteBuf buf) {
		super.writeInitial(buf);
		buf.writeByte(points.size());

		for (Vec2d point : points) ByteBufUtils.writeVec2d(buf, point);
	}

	@Override
	public void readInitial(@Nonnull PacketByteBuf buf) {
		super.readInitial(buf);
		int count = buf.readUnsignedByte();
		points.ensureCapacity(count);

		if (points.size() > count) points.subList(count, points.size()).clear();

		for (int i = 0; i < count; i++) {
			Vec2d point = ByteBufUtils.readVec2d(buf);
			if (i < points.size()) {
				points.set(i, point);
			} else {
				points.add(point);
			}
		}
	}

	@Override
	@Environment(EnvType.CLIENT)
	public void draw(@Nonnull CanvasClient canvas, @Nonnull MatrixStack matrices) {
		if (points.size() < 3) return;

		setupFlat();

		int size = points.size();
		Vec2d a = points.get(0);

		int red = getRed(), green = getGreen(), blue = getBlue(), alpha = getAlpha();

		BufferBuilder buffer = Tessellator.getInstance().getBuffer();
		Matrix4f matrix = matrices.peek().getPositionMatrix();

		buffer.begin(VertexFormat.DrawMode.TRIANGLES, VertexFormats.POSITION_COLOR);

		for (int i = 1; i < size - 1; i++) {
			Vec2d b = points.get(i), c = points.get(i + 1);
			buffer.vertex(matrix, (float) a.x(), (float) a.y(), 0).color(red, green, blue, alpha).next();
			buffer.vertex(matrix, (float) b.x(), (float) b.y(), 0).color(red, green, blue, alpha).next();
			buffer.vertex(matrix, (float) c.x(), (float) c.y(), 0).color(red, green, blue, alpha).next();
		}

		BufferRenderer.drawWithShader(buffer.end());
	}
}

File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/object2d/Polygon.kt
package pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d

import com.google.common.base.Objects
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.render.*
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.PacketByteBuf.getMaxValidator
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasClient
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ColourableObject
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectRegistry.POLYGON_2D
import pw.switchcraft.plethora.util.ByteBufUtils
import pw.switchcraft.plethora.util.Vec2d
import javax.annotation.Nonnull

open class Polygon(
  id: Int,
  parent: Int,
  type: Byte = POLYGON_2D
) : ColourableObject(id, parent, type), MultiPointResizable2d {
  protected var points = mutableListOf<Vec2d>()

  override fun setVertex(idx: Int, point: Vec2d) {
    if (!Objects.equal(points[idx], point)) {
      points[idx] = point
      setDirty()
    }
  }

  override val vertices: Int
    get() = points.size

  override fun getPoint(idx: Int): Vec2d = points[idx]

  override fun addPoint(idx: Int, @Nonnull point: Vec2d) {
    if (idx == points.size) {
      points.add(point)
    } else {
      points.add(idx, point)
    }
    setDirty()
  }

  override fun removePoint(idx: Int) {
    points.removeAt(idx)
    setDirty()
  }

  override fun writeInitial(buf: PacketByteBuf) {
    super.writeInitial(buf)
    buf.writeCollection(points, ByteBufUtils::writeVec2d)
  }

  override fun readInitial(buf: PacketByteBuf) {
    super.readInitial(buf)
    points = buf.readCollection(getMaxValidator({ mutableListOf() }, MAX_SIZE), ByteBufUtils::readVec2d)
  }

  @Environment(EnvType.CLIENT)
  override fun draw(canvas: CanvasClient, matrices: MatrixStack, consumers: VertexConsumerProvider?) {
    if (points.size < 3) return

    setupFlat()

    val size = points.size
    val a = points[0]

    val buffer = Tessellator.getInstance().buffer
    val matrix = matrices.peek().positionMatrix

    buffer.begin(VertexFormat.DrawMode.TRIANGLES, VertexFormats.POSITION_COLOR)

    for (i in 1 until size - 1) {
      val b = points[i]
      val c = points[i + 1]
      buffer.vertex(matrix, a.x.toFloat(), a.y.toFloat(), 0f).color(red, green, blue, alpha).next()
      buffer.vertex(matrix, b.x.toFloat(), b.y.toFloat(), 0f).color(red, green, blue, alpha).next()
      buffer.vertex(matrix, c.x.toFloat(), c.y.toFloat(), 0f).color(red, green, blue, alpha).next()
    }

    BufferRenderer.drawWithShader(buffer.end())
  }
}
